torch/csrc/distributed/rpc/rref_impl.cpp

```diff
@@ -167,40 +167,37 @@ IValue UserRRef::toHere() {
 //////////////////////////  OwnerRRef  /////////////////////////////////////
 
 const IValue& OwnerRRef::getValue() const {
-  std::unique_lock<std::mutex> lock(mutex_);
-  valueCV_.wait(lock, [this] { return value_.has_value(); });
-  return value_.value();
+  std::unique_lock<std::mutex> lock(futureValuePtrMutex_);
+  futureValueCompletedCV_.wait(
+      lock, [this] { return futureValuePtr_->completed(); });
+  return futureValuePtr_->wait();
 }
 
 bool OwnerRRef::hasValue() const {
-  std::lock_guard<std::mutex> lock(mutex_);
-  return value_.has_value();
+  std::lock_guard<std::mutex> lock(futureValuePtrMutex_);
+  return futureValuePtr_->completed();
 }
 
-std::shared_ptr<FutureMessage> OwnerRRef::getFuture() {
-  std::unique_lock<std::mutex> lock(mutex_);
-  if (future_.get()) {
-    return future_;
-  }
-  future_ = std::make_shared<FutureMessage>();
-  std::shared_ptr<FutureMessage> ret = future_;
-  if (value_.has_value()) {
-    lock.unlock();
-    ret->markCompleted(Message());
-  }
-  return ret;
+std::shared_ptr<torch::utils::Future<IValue>> OwnerRRef::getFuture() {
+  std::lock_guard<std::mutex> lock(futureValuePtrMutex_);
+  return futureValuePtr_;
 }
 
 void OwnerRRef::setValue(IValue&& value) {
-  std::unique_lock<std::mutex> lock(mutex_);
-  value_ = std::move(value);
-  std::shared_ptr<FutureMessage> future;
-  future.swap(future_);
-  lock.unlock();
-  valueCV_.notify_all();
-  if (future.get() && !future->completed()) {
-    future->markCompleted(Message());
+  std::unique_lock<std::mutex> lock(futureValuePtrMutex_);
+  if (futureValuePtr_->completed()) {
```

Comment:
The rest of this change seems fairly reasonable, but  this double-setValue() logic here seems to be potentially error-prone.

Do we need the ability to setValue() twice on an ownerRRef?

Creating a second future here seems to negate the value of having a Future<IValue> altogether. We can end up with problematic uses like the caller calling getFuture()->addCallback() and, if setValue is called twice, never being notified with the real value in some cases.

Reply:
> Do we need the ability to setValue() twice on an ownerRRef?

Based on the current design and API contract, yes, as RRef points to mutable data objects. 

I agree with @jjlilley's assessment here. When the future is already completed, `getFuture()` and the next `setValue()` could hit a race.

Reply:
It seems the only place that we call `RRef->getFuture` is in `request_callback_impl.cpp`, and purpose is to install a `callback` to the future. Will it be sufficient if we expose that `allCallback` API directly to RRef, make it atomic, and kill the `getFuture` API? Or is there any other potential use cases for `getFuture`?

Reply:
@jjlilley 

Yeah, it's like a dangling future, to which users can still add callbacks.

I think @mrshenli's solution is correct. This way, we hide future, and use OwnerRRef similarly to a future. Updated with this solution.

Reply:
It's usually a source of deadlocks to call something like markCompleted(), which may call callbacks, with the lock held.

The thing is, here, it's tricky to be correct here without the lock held.

Would we be better off instead just adding an explicit error+accessors to this class, rather than using Future<> in this regard?
It feels like we're trying to shoe-horn in the Future<IValue> here, with a pattern that doesn't quite fit.